java-frontend/src/main/java/org/sonar/java/se/checks/CustomUnclosedResourcesCheck.java

```diff
@@ -0,0 +1,453 @@
+        final ExpressionTree expression = syntaxNode.expression();
+        if (expression != null) {
+          if (expression.is(Tree.Kind.IDENTIFIER)) {
+            final IdentifierTree identifier = (IdentifierTree) expression;
+            currentVal = programState.getValue(identifier.symbol());
+          } else {
+            currentVal = programState.peekValue();
+          }
+          closeResource(currentVal);
+        }
+      }
+    }
+
+    private void closeArguments(final Arguments arguments, int stackOffset) {
+      final List<SymbolicValue> values = programState.peekValues(arguments.size() + stackOffset);
+      final List<SymbolicValue> argumentValues = values.subList(stackOffset, values.size());
+      for (SymbolicValue target : argumentValues) {
+        closeResource(target);
+      }
+    }
+  }
+
+  private class PostStatementVisitor extends AbstractStatementVisitor {
+
+    protected PostStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    @Override
+    public void visitNewClass(NewClassTree syntaxNode) {
+      if (isCreatingResource(syntaxNode)) {
+        final SymbolicValue instanceValue = programState.peekValue();
+        programState = programState.addConstraint(instanceValue, new ObjectConstraint(false, false, syntaxNode, Status.OPENED));
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxNode) {
+      if (isClosingResource(syntaxNode)) {
+        final ExpressionTree targetExpression = ((MemberSelectExpressionTree) syntaxNode.methodSelect()).expression();
+        if (targetExpression.is(Tree.Kind.IDENTIFIER)) {
+          final IdentifierTree identifier = (IdentifierTree) targetExpression;
+          closeResource(programState.getValue(identifier.symbol()));
+        } else {
+          closeResource(programState.peekValue());
+        }
+      } else if (isCreatingResource(syntaxNode)) {
+        final SymbolicValue instanceValue = programState.peekValue();
+        programState = programState.addConstraint(instanceValue, new ObjectConstraint(false, false, syntaxNode, Status.OPENED));
+      }
+    }
+
+    private boolean isCreatingResource(NewClassTree syntaxNode) {
+      for (ConstructorMatcher matcher : constructorClasses()) {
+        if (matcher.matches(syntaxNode)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    private boolean isCreatingResource(MethodInvocationTree syntaxNode) {
+      for (MethodMatcher matcher : factoryMethods()) {
+        if (matcher.matches(syntaxNode)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    private boolean isClosingResource(MethodInvocationTree methodInvocation) {
+      for (MethodMatcher matcher : closingMethods()) {
+        if (matcher.matches(methodInvocation)) {
+          return true;
+        }
+      }
+      return false;
+    }
+  }
+
+  private abstract static class ArgumentMatcher {
```

Comment:
I wonder why not get inspiration from disallowed method rule template ? https://github.com/SonarSource/sonar-java/blob/master/java-checks/src/main/java/org/sonar/java/checks/DisallowedMethodCheck.java 

That would simplify user experience and rule configuration. 
I understand that MethodMatcher classes are in java-check module but that maybe the time to change this.